synapse/lib/cell.py

```diff
@@ -2587,6 +2583,122 @@ async def _initCellBoot(self):
+
+            if os.path.isdir(fp):
+                logger.warning(f'Removing existing directory: {fp}')
+                shutil.rmtree(fp)
+                continue
+
+            logger.warning(f'Unhandled existing file: {fp}')  # pragma: no cover
+
+        # Setup get args
+        insecure_marker = 'https+insecure://'
+        kwargs = {}
+        if rurl.startswith(insecure_marker):
+            logger.warning(f'Disabling SSL verification for restore request.')
+            kwargs['ssl'] = False
+            rurl = 'https://' + rurl[len(insecure_marker):]
+
+        tmppath = s_common.gendir(dirn, 'tmp')
+        tarpath = s_common.genpath(tmppath, f'restore_{rurliden}.tgz')
+
+        try:
+
+            with s_common.genfile(tarpath) as fd:
+                async with aiohttp.client.ClientSession() as sess:
+                    async with sess.get(rurl, **kwargs) as resp:
+                        resp.raise_for_status()
+
+                        content_length = int(resp.headers.get('content-length', 0))
+                        if content_length > 100:
+                            logger.warning(f'Downloading {content_length/s_const.megabyte:.3f} MB of data.')
+                            pvals = [int((content_length * 0.01) * i) for i in range(1, 100)]
+                        else:  # pragma: no cover
+                            logger.warning(f'Odd content-length encountered: {content_length}')
+                            pvals = []
+
+                        csize = s_const.kibibyte * 64  # default read chunksize for ClientSession
+                        tsize = 0
+
+                        async for chunk in resp.content.iter_chunked(csize):
+                            fd.write(chunk)
+
+                            tsize = tsize + len(chunk)
+                            if pvals and tsize > pvals[0]:
+                                pvals.pop(0)
+                                percentage = (tsize / content_length) * 100
+                                logger.warning(f'Downloaded {tsize/s_const.megabyte:.3f} MB, {percentage:.3f}%')
+
+            logger.warning(f'Extracting {tarpath} to {dirn}')
+
+            with tarfile.open(tarpath) as tgz:
+                for memb in tgz.getmembers():
+                    if memb.name.find('/') == -1:
+                        continue
+                    memb.name = memb.name.split('/', 1)[1]
+                    logger.warning(f'Extracting {memb.name}')
+                    tgz.extract(memb, dirn)
+
+            # and record the rurliden
+            with s_common.genfile(donepath) as fd:
+                fd.truncate(0)
+                fd.seek(0)
+                fd.write(rurliden.encode())
+
+        except asyncio.CancelledError:  # pragma: no cover
+            raise
+
+        except Exception:  # pragma: no cover
+            logger.exception('Failed to restore cell from URL.')
+            raise
+
+        else:
+            logger.warning('Restored service from URL')
```

Comment:
Is there some kind of check we could do here to validate that we actually restored a cell vs extracting a random tarball?

Reply:
Could maybe check that there is a valid `cell.guid` file in the tarfile before extracting it